File: go-wasm/pkg/wasm/claimer.go
// +build wasm

package wasm

import (
	"encoding/json"
	"errors"
	"syscall/js"

	"github.com/KILTprotocol/portablegabi/go-wasm/pkg/credentials"
	"github.com/privacybydesign/gabi"
	"github.com/privacybydesign/gabi/revocation"
)

// GenKey creates the private key for the claimer
func GenKey(this js.Value, inputs []js.Value) (interface{}, error) {
	claimer, err := credentials.NewClaimer(SysParams)
	if err != nil {
		return nil, err
	}
	return claimer, nil
}

// KeyFromMnemonic derives a key from a given mnemonic
func KeyFromMnemonic(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 1 {
		return nil, errors.New("Missing mnemonic to generate claimer keys")
	}
	if len(inputs) < 2 {
		return nil, errors.New("Missing password to generate claimer keys")
	}
	claimer, err := credentials.ClaimerFromMnemonic(SysParams, inputs[0].String(), inputs[1].String())
	if err != nil {
		return nil, err
	}
	return claimer, nil
}

// RequestAttestation creates a session object and a message which request the
// attestation of specific attributes. The second object should be sent to an
// attester. This method expects as inputs the private key of the claimer, a
// json encoded string containing the claim which should be attested, the
// handshake message from the attester and the public key of the attester.
func RequestAttestation(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 4 {
		return nil, errors.New("Missing inputs to request attestation")
	}
	claimer := &credentials.Claimer{}
	claim := credentials.Claim{}
	handshakeMsg := &credentials.StartSessionMsg{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), &claim); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), handshakeMsg); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[3].String()), issuerPubKey); err != nil {
		return nil, err
	}

	session, msg, err := claimer.RequestAttestationForClaim(issuerPubKey, handshakeMsg, claim)
	if err != nil {
		return nil, err
	}
	return map[string]interface{}{
		"session": session,
		"message": msg,
	}, nil
}

// BuildCredential creates a credential which can be used to create proofs that
// the claimer posseses certain attributes. This method expects as input the
// private key of the claimer, the session object created by requestAttestation
// and the signature message send the attester.
func BuildCredential(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 3 {
		return nil, errors.New("Missing inputs to build credential")
	}
	claimer := &credentials.Claimer{}
	session := &credentials.UserIssuanceSession{}
	signature := &gabi.IssueSignatureMessage{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), session); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), signature); err != nil {
		return nil, err
	}

	credential, err := claimer.BuildCredential(signature, session)
	if err != nil {
		return nil, err
	}
	return credential, nil
}

// BuildPresentation creates a proof that the claimer posseses the requested
// attributes. This method takes as input the private key of the claimer, the
// credential which contains the requested attributes, a json encoded list
// containing the requested attributes and the public key of the attester.
// It returns a proof containing the values of the requested attributes.
func BuildPresentation(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 4 {
		return nil, errors.New("Missing inputs to build presentation")
	}
	claimer := &credentials.Claimer{}
	credential := &credentials.AttestedClaim{}
	request := &credentials.PresentationRequest{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), credential); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), request); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[3].String()), issuerPubKey); err != nil {
		return nil, err
	}

	disclosedAttr, err := claimer.BuildPresentation(issuerPubKey, credential, request)
	if err != nil {
		return nil, err
	}
	return disclosedAttr, nil
}

func BuildCombinedPresentation(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 4 {
		return nil, errors.New("Missing inputs to build combined presentation")
	}
	claimer := &credentials.Claimer{}
	creds := []*credentials.AttestedClaim{}
	request := &credentials.CombinedPresentationRequest{}
	attesterPubKey := []*gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), claimer); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), &creds); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), request); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[3].String()), &attesterPubKey); err != nil {
		return nil, err
	}

	disclosedAttr, err := claimer.BuildCombinedPresentation(attesterPubKey, creds, request)
	if err != nil {
		return nil, err
	}
	return disclosedAttr, nil
}

// UpdateCredential updates the non revocation witness using the provided update.
func UpdateCredential(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 3 {
		return nil, errors.New("Missing inputs to update credential")
	}
	credential := &credentials.AttestedClaim{}
	update := &revocation.Update{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), credential); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), update); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), issuerPubKey); err != nil {
		return nil, err
	}

	if err := credential.Update(issuerPubKey, update); err != nil {
		return nil, err
	}
	return credential, nil
}

// UpdateAllCredential updates the non revocation witness using all the provided updates.
func UpdateAllCredential(this js.Value, inputs []js.Value) (interface{}, error) {
	if len(inputs) < 3 {
		return nil, errors.New("Missing inputs to update credential")
	}
	credential := &credentials.AttestedClaim{}
	updates := []*revocation.Update{}
	issuerPubKey := &gabi.PublicKey{}

	if err := json.Unmarshal([]byte(inputs[0].String()), credential); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[1].String()), &updates); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(inputs[2].String()), issuerPubKey); err != nil {
		return nil, err
	}

	if err := credential.UpdateAll(issuerPubKey, updates); err != nil {
		return nil, err
	}
	return credential, nil
}

File: src/claim/Claimer.ts
import IClaimer, {
  AttestationRequest,
  ClaimerAttestationSession,
  Presentation,
  CombinedPresentation,
  ClaimError,
} from '../types/Claim'
import WasmHooks from '../wasm/WasmHooks'
import {
  IGabiMsgSession,
  InitiateAttestationRequest,
  Attestation,
  AttesterPublicKey,
} from '../types/Attestation'
import {
  CombinedPresentationRequest,
  PresentationRequest,
} from '../types/Verification'
import goWasmExec from '../wasm/wasm_exec_wrapper'
import Credential from './Credential'

/**
 * Checks that the provided claim is a valid object.
 *
 * @param claim The object which should be a valid claim.
 * @throws {ClaimError.claimMissing} If the claim is missing inside the [[AttestationRequest]].
 * @throws {ClaimError.notAnObject} If the [[Attestation]] object includes a non-object type claim.
 * @throws {ClaimError.duringParsing} If an error occurs during JSON deserialization.
 */
function checkValidClaimStructure(claim: object): void | Error {
  if (!Object.keys(claim).length) {
    throw ClaimError.claimMissing
  }
  if (typeof claim !== 'object') {
    throw ClaimError.notAnObject(typeof claim)
  }
  if (Array.isArray(claim)) {
    throw ClaimError.duringParsing
  }
}

export default class Claimer implements IClaimer {
  private readonly secret: string

  /**
   * Generates a claimer using the provided mnemonic.
   *
   * @param mnemonic The mnemonic which is used to generate the key.
   * @param password The password which is used to generate the key.
   * @returns A new claimer.
   */
  public static async buildFromMnemonic(
    mnemonic: string,
    password = ''
  ): Promise<Claimer> {
    // secret's structure unmarshalled is { MasterSecret: string }
    const secret = await goWasmExec<string>(WasmHooks.keyFromMnemonic, [
      mnemonic,
      password,
    ])
    return new this(secret)
  }

  /**
   * Generates a new master secret and returns a new [[Claimer]] object.
   *
   * @returns A new [[Claimer]].
   */
  public static async create(): Promise<Claimer> {
    const secret = await goWasmExec<string>(WasmHooks.genKey)
    return new this(secret)
  }

  /**
   * Constructs a new [[Claimer]] using the given master secret.
   *
   * @param secret The master secret of the [[Claimer]].
   */
  constructor(secret: string) {
    this.secret = secret
  }

  /**
   * Creates an [[AttestationRequest]] using the provided [[InitiateAttestationRequest]].
   *
   * @param p The parameter object.
   * @param p.claim The claim which should get attested.
   * @param p.startAttestationMsg The [[InitiateAttestationRequest]] provided by the attester.
   * @param p.attesterPubKey The [[PublicKey]] of the attester.
   * @returns An [[AttestationRequest]] and a [[ClaimerAttestationSession]] which together with an [[AttestationResponse]] can be used to create a [[Credential]].
   */
  public async requestAttestation({
    claim,
    startAttestationMsg,
    attesterPubKey,
  }: {
    claim: object
    startAttestationMsg: InitiateAttestationRequest
    attesterPubKey: AttesterPublicKey
  }): Promise<{
    message: AttestationRequest
    session: ClaimerAttestationSession
  }> {
    // check for invalid claim structure
    checkValidClaimStructure(claim)
    const { message, session } = await goWasmExec<IGabiMsgSession>(
      WasmHooks.requestAttestation,
      [
        this.secret,
        JSON.stringify(claim),
        startAttestationMsg.valueOf(),
        attesterPubKey.valueOf(),
      ]
    )
    return {
      message: new AttestationRequest(message),
      session: new ClaimerAttestationSession(session),
    }
  }

  /**
   * Builds a [[Credential]] using the [[ClaimerAttestationSession]] and the [[Attestation]].
   *
   * @param p The parameter object.
   * @param p.claimerSession The session object corresponding to the [[AttestationRequest]].
   * @param p.attestation The [[Attestation]] provided by the [[Attester]].
   * @returns A signed and valid [[Credential]].
   */
  public async buildCredential({
    claimerSession,
    attestation,
  }: {
    claimerSession: ClaimerAttestationSession
    attestation: Attestation
  }): Promise<Credential> {
    return new Credential(
      await goWasmExec<string>(WasmHooks.buildCredential, [
        this.secret,
        claimerSession.valueOf(),
        attestation.valueOf(),
      ])
    )
  }

  /**
   * Uses the [[PresentationRequest]] and a [[Credential]] to build an anonymous presentation.
   *
   * @param p The parameter object.
   * @param p.credential The [[Credential]] which contains all the requested attributes.
   * @param p.presentationReq The [[PresentationRequest]] received from the [[Verifier]].
   * @param p.attesterPubKey The public key of the [[Attester]] who signed the [[Credential]].
   * @returns A [[Presentation]] that can be used to disclose attributes with a [[Verifier]].
   *    Must only be used once!
   */
  public async buildPresentation({
    credential,
    presentationReq,
    attesterPubKey,
  }: {
    credential: Credential
    presentationReq: PresentationRequest
    attesterPubKey: AttesterPublicKey
  }): Promise<Presentation> {
    return new Presentation(
      await goWasmExec<string>(WasmHooks.buildPresentation, [
        this.secret,
        credential.valueOf(),
        presentationReq.valueOf(),
        attesterPubKey.valueOf(),
      ])
    )
  }

  /**
   * Uses the [[PresentationRequest]] and a [[Credential]] to build an anonymous presentation.
   *
   * @param p The parameter object.
   * @param p.credentials An array of [[Credential]]s which is used to provide the requested attributes.
   * @param p.combinedPresentationReq The array of [[PresentationRequest]]s received from the [[Verifier]].
   * @param p.attesterPubKeys An array of [[PublicKey]]s which corresponds to the array of [[Credential]]s.
   * @returns A [[CombinedPresentation]] that can be used to disclose attributes with a [[Verifier]].
   *    Must only be used once!
   */
  public async buildCombinedPresentation({
    credentials,
    combinedPresentationReq,
    attesterPubKeys,
  }: {
    credentials: Credential[]
    combinedPresentationReq: CombinedPresentationRequest
    attesterPubKeys: AttesterPublicKey[]
  }): Promise<CombinedPresentation> {
    // make a json array out of already json serialised values
    // we don't want a json array of strings
    return new CombinedPresentation(
      await goWasmExec<string>(WasmHooks.buildCombinedPresentation, [
        this.secret,
        `[${credentials.join(',')}]`,
        combinedPresentationReq.valueOf(),
        `[${attesterPubKeys.join(',')}]`,
      ])
    )
  }
}
